### PR TITLE
Unify client cert and cookie behaviour

### PR DESCRIPTION
**Pull request by jacobythwaites**
This PR allows a cookie session to be created (i.e. login) if you have a connection secured by client certificate. 

The current user for a REST call is now derived from the cookie session first, if present. If no cookie session, then the client certificate CN is used, if present. Otherwise, the call is associated with the unknown user.

There are several web UI changes as well to reflect this.

### CODE REVIEW

File: apps/sse_cfg/priv/gen_change.xsl [hunk 1]
```diff
@@ -21,14 +21,20 @@ inserting a change root element if necessary.
...
+  <xsl:strip-space elements='*'/>
+
+  <!--
+    The working directory is undefined in XSLT and is
+    passed in as a parameter.
+  -->
+  <xsl:param name='cwd' select='""'/>
 
   <!--
     The working directory is undefined in XSLT and is
```
Comment (lucafavatellawork) [inline apps/sse_cfg/priv/gen_change.xsl]: Duplicated. See original change in https://github.com/opensparkl/sse_core/pull/68/commits/b7bbac88db6588363bea9350ef271931a3191666#diff-1e8f8eee0f53132163454fd160258069 I expect you want to delete this instance.

File: apps/sse_cfg/priv/gen_change.xsl [hunk 2]
```diff
@@ -73,7 +79,31 @@ Note that the prop and grant elements are fields within a folder record.
...
+  <!--
+    Text elements are unwrapped and the text content inserted
+    directly.
+  -->
+  <xsl:template match='text'>
+    <xsl:value-of select='text()'/>
+  </xsl:template>
 
   <!--
     Includes are processed inline. Note that include hrefs are resolved
```
Comment (lucafavatellawork) [inline apps/sse_cfg/priv/gen_change.xsl]: Duplicated. See original change in https://github.com/opensparkl/sse_core/pull/68/commits/b7bbac88db6588363bea9350ef271931a3191666#diff-1e8f8eee0f53132163454fd160258069 I expect you want to delete this instance.

File: apps/sse_cfg/priv/gen_change.xsl [hunk 3]
```diff
@@ -73,7 +79,31 @@ Note that the prop and grant elements are fields within a folder record.
...
+        <xsl:value-of select='$cwd'/>
+        <xsl:text>/</xsl:text>
+      </xsl:if>
+      <xsl:value-of select='@href'/>
+    </xsl:variable>
+    <xsl:apply-templates select='document($pathname)/*'/>
+  </xsl:template>
+
+  <!--
+    Text elements are unwrapped and the text content inserted
```
Comment (lucafavatellawork) [inline apps/sse_cfg/priv/gen_change.xsl]: Duplicated. See original change in https://github.com/opensparkl/sse_core/commit/b7bbac88db6588363bea9350ef271931a3191666#diff-1e8f8eee0f53132163454fd160258069 I expect you want to delete other instance (line 130 i.e. 30 lines below).

**Comment by lucafavatellawork** [comment]
I found the discrepancies in file `apps/sse_cfg/priv/gen_change.xsl` by comparing this branch with the [(temporary) rebased branch `cert_nologin-rebased`](https://github.com/opensparkl/sse_core/commits/9f47a311afc3d1c0ff49662e7af7c65dc8b26f01) I created.

File: apps/sse_cfg/src/sse_cfg_configurator.erl [hunk 4]
```diff
@@ -92,22 +92,14 @@ configure({sse_cfg, crypto_algorithm}, undefined) ->
...
-  % The owner certificate can be claimed using the claim password
-  % provided by arg or env. If not supplied, the cert cannot be claimed.
-  OwnerCertClaim =
-    default_owner_claim(
-      env_owner_claim(
-        arg_owner_claim())),
-
   [
     % Owner must be TLS certificate authenticated. If the cert is to
     % be retrieved using the sse_cfg/claim REST call, the password must
```
Comment (lucafavatellawork) [inline apps/sse_cfg/src/sse_cfg_configurator.erl]:
```suggestion
    % be retrieved using the sse_cfg/cert REST call, the password must
```
?

Comment (lucafavatellawork) [reply apps/sse_cfg/src/sse_cfg_configurator.erl]: Not sure why GitHub shows as if I changed the number of spaces at the beginning of the line.

File: apps/sse_cfg/src/sse_cfg_yapp.erl [hunk 5]
```diff
@@ -99,30 +95,21 @@ handle_request(_Arg, 'DELETE',
...
 %% Returns the client certificate for the corresponding user, if
-%% the user exists, has a matching claim password, and has type SPARKL_CERT.
-%%
-%% If a user is defined but has password PASSWORD_NOT_APPLICABLE then
-%% the claim mechanism is not supported for that user.
+%% the user exists and the correct password is supplied.
 handle_request(Arg, 'POST',
-    "/claim", #?USER{} = User) ->
-  claim_cert(Arg, User);
+    "/cert", #?USER{} = User) ->
```
Comment (lucafavatellawork) [inline apps/sse_cfg/src/sse_cfg_yapp.erl]: I see the target `node_cert` in the makefile uses `/claim`. Shall that target be adapted or deleted?

File: apps/sse_cfg/src/sse_cfg_configurator.erl [hunk 6]
```diff
@@ -143,53 +135,3 @@ configure({sse_yaws, yaws_doc_roots}, Current) ->
...
-%% @doc
-%% Returns the CLAIM= value defined in the environment, or undefined
-%% if not present.
--spec
-env_owner_claim(undefined | string()) ->
-  undefined
-| string().
-
-env_owner_claim(undefined) ->
-  case os:getenv(?OWNER_CLAIM_ENV) of
```
Comment (lucafavatellawork) [inline apps/sse_cfg/src/sse_cfg_configurator.erl]: `-define(OWNER_CLAIM_ENV, "CLAIM").` above shall be deleted.

File: apps/sse_cfg/src/sse_cfg_configurator.erl [hunk 7]
```diff
@@ -143,53 +135,3 @@ configure({sse_yaws, yaws_doc_roots}, Current) ->
...
-
-%% @doc
-%% Returns the owner_claim arg value if defined, otherwise undefined.
--spec
-arg_owner_claim() ->
-  undefined
-| string().
-
-arg_owner_claim() ->
-  case init:get_argument(?OWNER_CLAIM_ARG) of
```
Comment (lucafavatellawork) [inline apps/sse_cfg/src/sse_cfg_configurator.erl]: `-define(OWNER_CLAIM_ARG, claim).` above shall be deleted.

File: apps/sse_cfg/src/sse_cfg_configurator.erl (hunk 6 above)
Comment (lucafavatellawork) [inline apps/sse_cfg/src/sse_cfg_configurator.erl]: The 3 definitions of this environment variable in the makefile (`CLAIM=postoffice`) shall be removed.

**Comment by lucafavatellawork** [comment]
I need to adapt further the sse_cli tests - in progress in https://github.com/opensparkl/sse_cli/pull/39

File: apps/sse_cfg/priv/gen_change.xsl (hunk 1 above)
Comment (lucafavatellawork) [reply apps/sse_cfg/priv/gen_change.xsl]: Addressed in #70

File: apps/sse_cfg/priv/gen_change.xsl (hunk 2 above)
Comment (lucafavatellawork) [reply apps/sse_cfg/priv/gen_change.xsl]: Addressed in #70

File: apps/sse_cfg/priv/gen_change.xsl (hunk 3 above)
Comment (lucafavatellawork) [reply apps/sse_cfg/priv/gen_change.xsl]: Addressed in #70

File: apps/sse_cfg/src/sse_cfg_configurator.erl (hunk 4 above)
Comment (lucafavatellawork) [reply apps/sse_cfg/src/sse_cfg_configurator.erl]: Addressed in #70

File: apps/sse_cfg/src/sse_cfg_yapp.erl (hunk 5 above)
Comment (lucafavatellawork) [reply apps/sse_cfg/src/sse_cfg_yapp.erl]: Addressed in #70

File: apps/sse_cfg/src/sse_cfg_configurator.erl (hunk 6 above)
Comment (lucafavatellawork) [reply apps/sse_cfg/src/sse_cfg_configurator.erl]: Addressed in #70

File: apps/sse_cfg/src/sse_cfg_configurator.erl (hunk 7 above)
Comment (lucafavatellawork) [reply apps/sse_cfg/src/sse_cfg_configurator.erl]: Addressed in #70

File: apps/sse_cfg/src/sse_cfg_configurator.erl (hunk 6 above)
Comment (lucafavatellawork) [reply apps/sse_cfg/src/sse_cfg_configurator.erl]: Addressed in #70

**Comment by lucafavatellawork** [comment]
> I need to adapt further the sse_cli tests - in progress in [opensparkl/sse_cli#39](https://github.com/opensparkl/sse_cli/pull/39)

Done.

File: apps/sse_yaws/src/sse_yaws_util.erl [hunk 8]
```diff
@@ -502,41 +510,94 @@ delete_cookie_session(Arg) ->
...
+          {ok, sse_cfg_prop:set(User, Prop)};
+
+        false ->
+          {ok, ?UNKNOWN_USER}
+      end
+  end.
+
+%% @doc
+%% Gets the user associated with the current session cookie.
+%% If no session is established, the unknown user is returned.
```
Comment (lucafavatellawork) [inline apps/sse_yaws/src/sse_yaws_util.erl]:
```suggestion
%% If no session is established, an error is returned.
```

File: apps/sse_yaws/src/sse_yaws_util.erl [hunk 9]
```diff
@@ -502,41 +510,94 @@ delete_cookie_session(Arg) ->
...
 
 %% @doc
-%% Gets the user associated with the session, or the constant unknown
-%% user if no user is in session.
+%% Gets the user associated with the cookie session. If none,
+%% and the connection is secured by client certificate, gets
+%% the cert user instead. Otherwise, returns the constant
+%% unknown user.
+%%
+%% If not the unknownn user, the user_type prop is set to
```
Comment (lucafavatellawork) [inline apps/sse_yaws/src/sse_yaws_util.erl]:
```suggestion
%% If not the unknown user, the user_type prop is set to
```

File: apps/sse_cfg/priv/docroot/account/account.js [hunk 10]
```diff
@@ -16,60 +16,101 @@
...
+const CERT_PATH = '/sse_cfg/cert'
 const RENDER_PAGE = '/cfg/render/render.html'
+const UNKNOWN_USER = 'unknown@unknowndomain'
+
+const LISTENERS = 'listeners'
+const OPENERS = 'openers'
 
 /**
- * Called on load. Initialises the button events.
+ * Called on load. Thie wires up events and gets the user, which
```
Comment (lucafavatellawork) [inline apps/sse_cfg/priv/docroot/account/account.js]:
```suggestion
 * Called on load. This wires up events and gets the user, which
```
?

File: apps/sse_cfg/src/sse_cfg_yapp.erl [hunk 11]
```diff
@@ -345,6 +249,53 @@ assert_not_self(#?USER{name=Caller}, #?USER{name=Caller}) ->
...
+%% If the correct claim password is given, responds with a new
+%% combo client certificate for the requested user name.
+%%
+%% The combo file contains both the client key and cert signed by the
+%% node's self-signed root CA.
+%%
+%% The user of type SPARKL_CERT must already exist with a matching claim
+%% password, otherwise an error response is given.
+%%
+%% The receiver can convert the PEM into PKS#12 using the openssl
```
Comment (lucafavatellawork) [inline apps/sse_cfg/src/sse_cfg_yapp.erl]:
```suggestion
%% The receiver can convert the PEM into PKCS#12 using the openssl
```

File: apps/sse_cfg/src/sse_cfg_yapp.erl [hunk 12]
```diff
@@ -345,6 +249,53 @@ assert_not_self(#?USER{name=Caller}, #?USER{name=Caller}) ->
 assert_not_self(_Caller, _DeleteUser) ->
   ok.
 
+%% @doc
+%% If the correct claim password is given, responds with a new
```
Comment (lucafavatellawork) [inline apps/sse_cfg/src/sse_cfg_yapp.erl]:
```suggestion
%% If the correct password is given, responds with a new
```

File: apps/sse_cfg/src/sse_cfg_yapp.erl [hunk 13]
```diff
@@ -345,6 +249,53 @@ assert_not_self(#?USER{name=Caller}, #?USER{name=Caller}) ->
...
+%% combo client certificate for the requested user name.
+%%
+%% The combo file contains both the client key and cert signed by the
+%% node's self-signed root CA.
+%%
+%% The user of type SPARKL_CERT must already exist with a matching claim
```
Comment (lucafavatellawork) [inline apps/sse_cfg/src/sse_cfg_yapp.erl]:
```suggestion
%% The user must already exist with a matching
```

File: apps/sse_cfg/priv/docroot/account/account.js (hunk 10 above)
Comment (jacobythwaites) [reply apps/sse_cfg/priv/docroot/account/account.js]: Fixed

**Pull request merged**